### PR TITLE
39-implement article state transitions

### PR DESCRIPTION
**Pull request by isuriupeksha**
1. By default articles can be saved draft when user is `contributor`
2. `Admin` and `Volunteer` can save articles as draft or published.

### CODE REVIEW

File: app/controllers/articles_controller.rb [hunk 1]
```diff
@@ -48,10 +50,44 @@ def update
       render 'edit'
     end
   end
+
+  def publish
+    respond_with(@article) do |format|
+      if @article.publish!
+        flash[:notice] = "Article has been publish."
```
Comment (wonook) [inline app/controllers/articles_controller.rb]: The article has been published

File: app/controllers/articles_controller.rb [hunk 2]
```diff
@@ -48,10 +50,44 @@ def update
...
+        format.html { redirect_to article_path(@article) }
+        format.json { render json: @article.errors.to_hash(true), status: :unprocessable_entity }
+      end
+    end
+  end
+
+  def unpublish
+    respond_with(@article) do |format|
+      if @article.unpublish!
+        flash[:notice] = "Article has been unpublish."
```
Comment (wonook) [inline app/controllers/articles_controller.rb]: The article has been unpublished

File: app/controllers/articles_controller.rb [hunk 3]
```diff
@@ -48,10 +50,44 @@ def update
...
+
+        format.html { redirect_to article_path(@article) }
+        format.json { render json: @article, location: article_path(@article) }
+      else
+        flash[:error] = "Sorry, failed to to unpublish the article"
```
Comment (wonook) [inline app/controllers/articles_controller.rb]: Sorry, failed to to unpublish the article --> Failed to draft the article, please try again

File: db/migrate/20150713020715_add_state_column_to_articles.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+class AddStateColumnToArticles < ActiveRecord::Migration
+  def change
+    add_column :articles, :state, :string, default: "draft"
```
Comment (wonook) [inline db/migrate/20150713020715_add_state_column_to_articles.rb]: I thought we went over this in our discussions on #39 . Please read over what @abelards has mentioned. I don't think this does more than anything more than a `bool` but just taking up more space.

**Comment by wonook** [comment]
By the way, I have a feeling that using `aasm` gem for just this is a bit too much. We only have to have two states for each articles, and that is either 'published' or 'drafted'. Why not implement this in the manner I have implemented the `Login Approval` thing in the `User` model? Please refer to #46 .

**Comment by abelards** [comment]
Agreed with @wonook on `AASM` being overkill.
Timestamps are great as a "poor man's state machine", and often just enough.

I acknowledge that big projects often understand the need of a state machine too late, it's good to have thought about the question. However it would have been nice to talk about it beforehand, we would have asked @rosariorobinson on a few more possible states (confirmed, obsolete, and so on).
And as always, changing scope in mid-project looks nice but might make you late so perhaps for everyone's best interests we would have agreed on which features get in, and which get out.

On a side note, if @Isuri wants to learn and practice state machines, well, that's also a project for her to learn and I am OK with that. I don't consider having AASM as a blocker :)

**Comment by isuriupeksha** [comment]
@wonook @abelards Thanks for the feedback. But here I have implemented this, not only two state. It may have more state when app grow. I am planing to add transaction activities as well. 
code sample: `on_transition: Proc.new { |obj, *args| obj.record_activity(*args); }` (it has nice features)

When consider @wonook suggested way, we can have only two state. But we cannot expect this article transition has only two. With this implementation any future developer can add a additional `state` without changing the database and more complex way to edit model.

**Comment by isuriupeksha** [comment]
Fixed typo issues.

**Comment by wonook** [comment]
I think it would make more sense if you could give an example of a case where these state transitions can be helpful. I still think that `aasm` is an overkill for this case. We can mark an article's state with a simple `int` if not `string` or `bool` or `timestamp`, and that would be enough *and* more flexible with state transitions as well. I think that this would seem reasonable once we have *a situation where we want to go from one state to the other and not the other way around*. It's another burden for others later on taking these stuff out if they decide that they don't need such functionalities.
Maybe it would make more sense if you could share an example of a simple state diagram or anything similar for an `article`'s lifecycle that you have in your mind.

**Comment by wonook** [comment]
I'll merge this one into our develop branch as well at around 4-5AM GMT+0. Don't worry about the merge conflicts I'll take care of them. I hope that @isuriupeksha would take full responsibility for this PR, as she insists to use the `aasm` gem for any kind of consequences. I hope that future developers would be able to get in touch with her for any of the questions regarding this PR.

**Pull request merged**